GAWB-4026: Blocking, Scalike, Liquibase, Http4s Routes, Config Rendering + some refactoring

ansingh7115 opened the pull request:
This PR contains:
- a refactor of hamm per the team's decision to use blocking architecture.
- classes for Scalike 
- liquibase for db migration management 
- Http4s for routing

Qi77Qi commented on core/src/main/scala/org/broadinstitute/workbench/hamm/dao/GooglePriceListDAO.scala:
```diff
@@ -13,34 +13,33 @@ import org.http4s.client.Client
+    val key = "[KEY]"
     for {
-      googlePriceList <- httpClient.expect[GooglePriceList](uri)
+      googlePriceList <- httpClient.expect[GooglePriceList](uri + s"/v1/services/$serviceId/skus?key=$key")
```
is this api key that's needed for authenticating against google pricing api? if so, it probably should be pulled from vault?

Qi77Qi commented on core/src/main/scala/org/broadinstitute/workbench/hamm/model/model.scala:
```diff
@@ -414,8 +424,16 @@ final case class Call(runtimeAttributes: RuntimeAttributes,
+  def apply(calls: List[Call], startTime: Instant, endTime: Instant, labels: Map[String, String]): MetadataResponse = {
+    labels.get("caas-collection-name") match {
+      case Some(c) => MetadataResponse(calls, startTime, endTime, WorkflowCollectionId(UUID.fromString(c)), labels)
+      case None => throw new Exception(s"Workflow did not have a collection associated with it.")
```
this seems kinda weird pattern since the case class already gives an `apply` method. What's the gain with overloading `apply`?

 If you need to do this labels check, can you just do it in https://github.com/DataBiosphere/hamm/blob/master/core/src/main/scala/org/broadinstitute/workbench/hamm/model/CromwellMetadataJsonCodec.scala#L87 ? That way, you don't need to throw an exception either.

Qi77Qi commented on server/src/main/scala/org/broadinstitute/workbench/hamm/server/WorkflowCostService.scala:
```diff
@@ -1,50 +1,73 @@
+                                              samDAO: HttpSamDAO[F]) extends HammFs2Grpc[F] {
+
   override def getWorkflowCost(request: WorkflowCostRequest, clientHeaders: Metadata): F[WorkflowCostResponse] = {
+    withAuthenticatedUser(clientHeaders) { userInfo =>
```
is this authenticate or authorize?

Qi77Qi commented on server/src/main/scala/org/broadinstitute/workbench/hamm/server/WorkflowCostService.scala:
```diff
@@ -1,50 +1,73 @@
+    withAuthenticatedUser(clientHeaders) { userInfo =>
+      //ToDo: some work here to make this less messy
+      for {
+        cromwellMetadata: MetadataResponse <- workflowDAO.getMetadata(WorkflowId(UUID.fromString(request.id)))
```
if `: MetadataResponse` is not needed, it's better to remove it...cuz it'll add an additional function call when things are desugared

Qi77Qi commented on core/src/main/scala/org/broadinstitute/workbench/hamm/dao/GooglePriceListDAO.scala:
```diff
@@ -78,7 +87,10 @@ object GooglePriceListDAO {
-    } yield PriceList(computePricesParseq, storagePricesParseq)
+    } yield {
+      val priceList = PriceList(computePricesParseq, storagePricesParseq)
+      priceList
```
is this change intended?

Qi77Qi commented on core/src/main/scala/org/broadinstitute/workbench/hamm/dao/WorkflowMetadataDAO.scala:
```diff
@@ -3,11 +3,12 @@ package org.broadinstitute.workbench.hamm.dao
+class WorkflowMetadataDAO[F[_]: Sync](httpClient: Client[F], uri: Uri) {
   def getMetadata(workflowId: WorkflowId): F[MetadataResponse] = {
-    httpClient.expect[MetadataResponse](s"https://cromwell.dsde-alpha.broadinstitute.org/api/workflows/v1/${workflowId.uuid}/metadata") //TODO: figure out how to get the url properly
+    httpClient.expect[MetadataResponse](uri + s"/${workflowId.uuid}/metadata") //TODO: figure out how to get the url properly
```
I'd prefer to only move the host name to config and leave the `/api/workflows/v1/` part here just like in other places

Qi77Qi commented on core/src/main/scala/org/broadinstitute/workbench/hamm/dao/WorkflowMetadataDAO.scala:
```diff
@@ -3,11 +3,12 @@ package org.broadinstitute.workbench.hamm.dao
 import org.http4s.client.Client
 
-class WorkflowMetadataDAO[F[_]: Sync](httpClient: Client[F]) {
+class WorkflowMetadataDAO[F[_]: Sync](httpClient: Client[F], uri: Uri) {
```
can you rename `uri` to something like `cromwellHost`

ansingh7115 commented on application.conf.ctmpl:
```diff
@@ -0,0 +1,44 @@
+    poolConnectionTimeoutMillis = 5000
+    driver = "org.postgresql.Driver"
+  }
+}
```
This is how scalike likes it's configs to look. Scalike supports Typesafe configs, which is used across all our repos, which is why I switched us back to using that instead of pureconfig

Qi77Qi replied:
pureconfig supports both typesafe config and camel case(https://github.com/pureconfig/pureconfig/issues/239#issuecomment-303574235). You wouldn't need to define those implicits manually except for some more custom ones. It'd also write unit tests for configurations

ansingh7115 replied:
I put those config items there because I think those are the ones we might end up tweaking. I think we should still use this config tool because that's what we use across our other repos and I don't see much of a reason to change it to pureconfig. We also haven't ever tested our configs

Qi77Qi replied:
I don't think never testing configs is a good tradition.

ansingh7115 replied:
Okay! We can decide that later :)

ansingh7115 commented on core/src/main/resources/org/broadinstitute/dsde/workbench/hamm/liquibase/changesets/20190225_create_tables.xml:
```diff
@@ -0,0 +1,71 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
These are the changesets for creating the `Job` and `Workflow` table and any unique or foreign keys. I took these from the doobie table definitions. I added a foreign key on `workflowId` because every job should belong to a workflow.

Questions:
- The naming for the job table columns is confusing. What is `CALL_FQN`, `ATTEMPT`, "VENDOR_JOB_ID" and `JOB_INDEX`? 
- Question from Doug - Is `IS_SUBWORKFLOW` necessary?

ansingh7115 replied:
1. Good point - but in order to tell a user how much a particular job/task cost, we need to be able to check against a workflow collection in sam, which means that the job needs to be associated with a workflow in order for us to authenticate. 

2. No, the operation id cost idea was based on a hunch by someone and it wasn't a real thing :( . I still don't have a clear understanding of what CALL_FQN and ATTEMPT mean - can you explain what they are? 

3.  I can keep `IS_SUBWORKFLOW` for now, we can delete it later if we don't need it

Qi77Qi replied:
`CALL_FQN` is just a name for the job(but it's not unique across workflow, hence you need the combined key), `attempt` is how many times this job has been retried.

geoffjentry replied:
@ansingh7115  All of those but `VENDOR_JOB_ID` come from Cromwell

ansingh7115 replied:
So in cromwell metadata that starts like this: 
` """
      |{
      |  "actualWorkflowLanguage": "WDL",
      |  "actualWorkflowLanguageVersion": "draft-2",
      |  "calls": {
      |    "echo_strings.echo_files": [
      |      {
      |        "attempt": 1,
      |        "backend": "JES",
      |        "backendLabels": {
      |          "cromwell-workflow-id": "cromwell-0942e6dc-2a2e-4912-86e2-b91fdaf06c44",
      |          "wdl-task-name": "echo-files"
      |        },
      |        "backendLogs": {
      |          "log": "gs://fc-d28aeae4-30cc-46ca-986d-2ffda410f47c/e9e9d8ef-7ca0-4c69-abc2-e93e959c81a8/echo_strings/c5b6ee46-1f09-4830-91a8-fd814866d664/call-echo_files/echo_files.log"
      |        },`

the `CALL_FQN` would be "echo_strings.echo_files"? I'm just trying to pinpoint exactly what it is so I know where it's coming from

ansingh7115 commented on core/src/main/scala/org/broadinstitute/workbench/hamm/HammLogger.scala:
```diff
@@ -0,0 +1,9 @@
+package org.broadinstitute.workbench.hamm
+
+import org.log4s._
```
Using log4s for logging

ansingh7115 commented on core/src/main/scala/org/broadinstitute/workbench/hamm/db/DbReference.scala:
```diff
@@ -0,0 +1,79 @@
+package org.broadinstitute.workbench.hamm.db
```
DbReference is adapted to use Scalike

ansingh7115 commented on core/src/main/scala/org/broadinstitute/workbench/hamm/db/DbReference.scala:
```diff
@@ -0,0 +1,79 @@
+    DB.localTx[A] { implicit session =>
+      f(session)
+    }
+  }
```
`inReadOnlyTransaction` and `inLocalTransaction` are the two transaction types. The former is the only type of transaction that the request side of hamm will be using.

ansingh7115 commented on core/src/main/scala/org/broadinstitute/workbench/hamm/db/JobTableQueries.scala:
```diff
@@ -0,0 +1,111 @@
+package org.broadinstitute.workbench.hamm.db
```
Scalike queries for the Job table: 
- insert job
- get job
- get job cost
- get job's workflow's collection id   (<--- why I added a foreign key on the workflow table)

ansingh7115 commented on core/src/main/scala/org/broadinstitute/workbench/hamm/db/WorkflowTableQueries.scala:
```diff
@@ -0,0 +1,130 @@
+import org.broadinstitute.workbench.hamm.model._
+import org.postgresql.util.PGobject
+
+
```
Scalike queries for the Workflow table:
- insert workflow
- get workflow
- get workflow cost

The `LabelTypeBinder` and and `LabelPbf` are binders for jsonb

ansingh7115 commented on core/src/main/scala/org/broadinstitute/workbench/hamm/api/HammRoutes.scala:
```diff
@@ -0,0 +1,81 @@
+import org.http4s._
+import org.http4s.server.middleware.Logger
+import org.http4s.syntax.kleisli._
+
```
Http4s routes class. Each call to the service is wrapped in an IO. 
I just replaced all the `F`s with `IO`s since our use of it is limited and we're really never going to use anything but IO there.

ansingh7115 commented on core/src/main/scala/org/broadinstitute/workbench/hamm/auth/SamSwaggerClient.scala:
```diff
@@ -0,0 +1,27 @@
+package org.broadinstitute.workbench.hamm.auth
```
We're hitting sam with the published swagger client.

ansingh7115 commented on core/src/main/scala/org/broadinstitute/dsde/workbench/hamm/HammLogger.scala:
```diff
@@ -0,0 +1,9 @@
+package org.broadinstitute.dsde.workbench.hamm
+
+import org.log4s._
```
Using log4s for logging

ansingh7115 commented on core/src/main/scala/org/broadinstitute/dsde/workbench/hamm/db/DbReference.scala:
```diff
@@ -0,0 +1,79 @@
+import org.broadinstitute.dsde.workbench.hamm.HammLogger
+import sun.security.provider.certpath.SunCertPathBuilderException
+import scalikejdbc._
+
```
DbReference is adapted to use Scalike

ansingh7115 commented on core/src/main/scala/org/broadinstitute/dsde/workbench/hamm/db/DbReference.scala:
```diff
@@ -0,0 +1,79 @@
+      f(session)
+    }
+  }
+}
```
inReadOnlyTransaction and inLocalTransaction are the two transaction types. The former is the only type of transaction that the request side of hamm will be using.

ansingh7115 commented on core/src/main/scala/org/broadinstitute/dsde/workbench/hamm/db/JobTableQueries.scala:
```diff
@@ -0,0 +1,111 @@
+import org.broadinstitute.dsde.workbench.hamm.HammLogger
+import org.broadinstitute.dsde.workbench.hamm.model._
+import scalikejdbc._
+
```
Scalike queries for the Job table:

- insert job
- get job
- get job cost
- get job's workflow's collection id (<--- why I added a foreign key on workflowId)

ansingh7115 commented on core/src/main/scala/org/broadinstitute/dsde/workbench/hamm/db/WorkflowTableQueries.scala:
```diff
@@ -0,0 +1,130 @@
+import org.broadinstitute.dsde.workbench.hamm.model._
+import org.postgresql.util.PGobject
+
+
```
Scalike queries for the Workflow table:

- insert workflow
- get workflow
- get workflow cost

The LabelTypeBinder and and LabelPbf are binders for jsonb

ansingh7115 commented on core/src/main/scala/org/broadinstitute/dsde/workbench/hamm/api/HammRoutes.scala:
```diff
@@ -0,0 +1,81 @@
+import org.http4s._
+import org.http4s.server.middleware.Logger
+import org.http4s.syntax.kleisli._
+
```
Http4s routes class. Each call to the service is wrapped in an IO.
I just replaced all the Fs with IOs since our use of it is limited and we're really never going to use anything but IO there.

ansingh7115 commented on core/src/main/scala/org/broadinstitute/dsde/workbench/hamm/auth/SamSwaggerClient.scala:
```diff
@@ -0,0 +1,27 @@
+import org.broadinstitute.dsde.workbench.hamm.model.{SamResource, SamResourceAction, SamResourceType}
+import org.http4s.Credentials.Token
+import org.http4s.Uri
+
```
We're hitting sam with the published swagger client.

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/workbench/hamm/api/HammRoutes.scala:
```diff
@@ -0,0 +1,81 @@
+
+
+  private def extractToken(request: Request[IO]): Token = {
+    val unauthorizedException = HammException(Status.Unauthorized.code, "User is unauthorized.")
```
why have this as a val here?

dvoet replied:
make it lazy

dvoet commented on project/Dependencies.scala:
```diff
@@ -1,28 +1,43 @@
+  val liquibaseVersion      = "3.5.3"
+  val minitestVersion       = "2.2.2"
+  val postgresDriverVersion = "42.2.4"
+  val samV                  = "0.1-b87909e-SNAP"
```
update with non-snap version

ansingh7115 commented on core/src/main/scala/org/broadinstitute/dsde/workbench/hamm/dao/GooglePriceListDAO.scala:
```diff
@@ -0,0 +1,80 @@
+package org.broadinstitute.dsde.workbench.hamm.dao
+
```
No changes here, just a package rename

ansingh7115 commented on core/src/main/scala/org/broadinstitute/dsde/workbench/hamm/dao/WorkflowMetadataDAO.scala:
```diff
@@ -0,0 +1,46 @@
+package org.broadinstitute.dsde.workbench.hamm.dao
+
+
```
No changes here, just a package rename

ansingh7115 commented on core/src/test/scala/org/broadinstitute/dsde/workbench/hamm/api/HammRoutesSpec.scala:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    check(response, Status.Ok, Some(WorkflowCostResponse(TestData.testWorkflowId, TestData.testWorkflow.cost)))
+  }
```
This test is failing, I think because we're starting a `session` here but the CostService is also starting a session?

ansingh7115 commented on core/src/test/scala/org/broadinstitute/dsde/workbench/hamm/db/JobTableQueriesSpec.scala:
```diff
@@ -0,0 +1,39 @@
+package org.broadinstitute.dsde.workbench.hamm.db
+
+import java.time.Instant
+
```
Switched us back to using `scalatest` because scalike supports it and gives us a (sometimes) handy `AutoRollback` trait.

Qi77Qi commented:
1. I'm aware team decided to go with `blocking`, but I don’t think that should mean `unsafeRunSync` everywhere…especially for http requests, which is built async, and we're blocking to get the result, and then rewrapping it in IO, which is pretty bad.

2. Don't put the http server in `core` module, which means `cost-Updater` will include the api server as well, and we definitely don’t want to do that

3. This PR is way too big…it’ll be much more effective review if you can separate things out as apposed to keep adding more changes to the same PR. I would do all these things in different PRs, 1. replace doobie with scalike (even though I would really prefer doobie) 2. http routes 3. liquidbase 4. Add swagger 5. authorization with sam.

ansingh7115 commented:
@Qi77Qi 

1. unsafeRunSync is currently being used only in getting the price list from google. I deleted the Cromwell call to get Metadata, per Doug's request. And anywhere else we use unsafeRunSync is in tests. For the call to google, I'm changing it in my next PR to use the cloudcalculator price list instead, so I think we can change the unsafeRunSync to something better in that PR.

2. Added back the server package

3. You're right, I'll try to be more strict about that next time. :)

Qi77Qi commented on server/src/main/scala/org/broadinstitute/dsde/workbench/hamm/Main.scala:
```diff
@@ -0,0 +1,49 @@
+      statusService       = new StatusService
+      hammRoutes          = new HammRoutes(samAuthProvider, costService, statusService)
+      routes              = hammRoutes.routes
+      server              <- BlazeServerBuilder[IO].bindHttp(8080, "localhost").withHttpApp(routes).serve    //.compile.drain.as(ExitCode.Success)  // new WorkflowCostService[IO](pricing, metadataDAO, samAuthProvider)).start
```
can you remove the commented out lines

Qi77Qi commented on server/src/main/scala/org/broadinstitute/dsde/workbench/hamm/api/HammRoutes.scala:
```diff
@@ -0,0 +1,79 @@
+    case GET -> Root =>
+      Ok( IO { statusService.status() } )
+    case _ -> Root =>
+      MethodNotAllowed(Allow(GET))
```
is this needed? you've already defined `orNotFound` above

Qi77Qi commented on server/src/main/scala/org/broadinstitute/dsde/workbench/hamm/api/HammRoutes.scala:
```diff
@@ -0,0 +1,79 @@
+    case request @ GET -> Root / "job" / jobId =>
+      Ok(IO { costService.getJobCost(extractToken(request), JobId(jobId)) })
+    case _ -> Root =>
+      MethodNotAllowed(Allow(GET))
```
same comment

Qi77Qi commented on server/src/main/scala/org/broadinstitute/dsde/workbench/hamm/api/HammRoutes.scala:
```diff
@@ -0,0 +1,79 @@
+
+
+
+  private def extractToken(request: Request[IO]): Token = {
```
I would make this return an `Either[Throwable, Token]`, and at call site such as https://github.com/DataBiosphere/hamm/pull/5/files#diff-e77976a07ebc1116eacf55f51ac2b49bR38 something like

```
      for {
        token  <- IO.fromEither(extractToken(request))
        cost <- IO { costService.getWorkflowCost(token, WorkflowId(workflowId)) }
        resp <- Ok(cost)
      } yield resp
```

You don't have to squeezing everything in one line and wrap it in `OK`

ansingh7115 replied:
this feels like more code each time we make an authenticated call

Qi77Qi replied:
but it's a lot easier to see what could fail and what the errors are. By looking at `costService.getWorkflowCost(extractToken(request), WorkflowId(workflowId))`, you'll have to look closer to realize there's an `extractToken` call that could fail.

ansingh7115 replied:
I thought that it might be too much repetitive code, so I tried a different thing and added auth middleware

ansingh7115 commented on core/src/main/scala/org/broadinstitute/dsde/workbench/hamm/service/StatusService.scala:
```diff
@@ -0,0 +1,12 @@
+package org.broadinstitute.dsde.workbench.hamm.service
+
+
+class StatusService  {
```
We eventually want to capture the health of not only hamm but services it depends on (such as sam)

ansingh7115 commented on core/src/test/scala/org/broadinstitute/dsde/workbench/hamm/TestComponent.scala:
```diff
@@ -0,0 +1,45 @@
+import scala.concurrent.ExecutionContext
+
+
+trait TestComponent extends Matchers with ScalaFutures {
```
copy paste :)

ansingh7115 commented on cost-updater/src/main/scala/org/broadinstitute/dsde/workbench/hamm/costUpdater/Main.scala:
```diff
@@ -0,0 +1,62 @@
+//package org.broadinstitute.dsde.workbench.hamm
```
I know you have a PR for this, so I just commented it out for now so I wouldn't have to bother with it

Qi77Qi commented on build.sbt:
```diff
@@ -1,4 +1,4 @@
-coverageExcludedPackages := "org.broadinstitute.workbench.hamm.protos"
+coverageExcludedPackages := "org.broadinstitute.dsde.workbench.hamm.protos"
```
I thought `dsde` is an outdated name, shouldn't we leave it out?

ansingh7115 replied:
I thought it was supposed to be consistent with everywhere else - @doge do you know if it would be a big deal if we left it out?

ansingh7115 replied:
changed it to `dsp`

ansingh7115 commented on core/src/test/scala/org/broadinstitute/dsde/workbench/hamm/TestExecutionContext.scala:
```diff
@@ -0,0 +1,20 @@
+
+import scala.concurrent.ExecutionContextExecutor
+
+object TestExecutionContext {
```
It's used in `DbSingleton` to create an execution context for our tests and also in `TestComponent`

ansingh7115 commented on automation/src/test/scala/org/broadinstitute/dsde/workbench/hamm/automation/WorkflowCostTest.scala:
```diff
@@ -1,11 +1,11 @@
-//package org.broadinstitute.workbench.hamm
+//package org.broadinstitute.dsde.workbench.hamm
```
I added a ticket to add automation - I think we'll have to talk about how best to do it since we aren't using GRPC anymore - https://broadworkbench.atlassian.net/browse/AP-151

Qi77Qi replied:
I think the only difference from what it was before is how we call the APIs. So instead of calling hamm api with grpc client, we're using a http client

ansingh7115 replied:
This PR is already really long and I don't want to add more to it to review - I would like to merge it and get out of this prolonged review cycle

Qi77Qi replied:
I think the fact that this PR got so many things piled together is the reason why review is hard and it takes this long, which I mentioned in the other comment.

It's not ideal that we're deleting tests and not adding them back, but I'm fine with it being added in a followup PR at this point.

Qi77Qi commented on automation/src/test/scala/org/broadinstitute/dsp/workbench/hamm/automation/AutomationTestConfig.scala:
```diff
@@ -1,10 +1,9 @@
-package org.broadinstitute.dsde.workbench.hamm
-package automation
+package org.broadinstitute.dsp.workbench.hamm.automation
```
the previous way of importing will import things in core package, which is usually what we want.

Qi77Qi commented on core/src/main/scala/org/broadinstitute/dsp/workbench/hamm/db/JobTable.scala:
```diff
@@ -106,16 +106,16 @@ object JobBinders {
-  implicit val callFqnTypeBinder: TypeBinder[CallFqn] = new TypeBinder[CallFqn] {
-    def apply(rs: ResultSet, label: String): CallFqn = CallFqn(rs.getString(label))
-    def apply(rs: ResultSet, index: Int): CallFqn = CallFqn(rs.getString(index))
+  implicit val callNameTypeBinder: TypeBinder[CallName] = new TypeBinder[CallName] {
```
I don't think it's a good idea to name this `CallName`. Like Jeff said, it's from cromwell's metadata, I think it's less confusing to call it the same way

Qi77Qi commented on core/src/main/scala/org/broadinstitute/dsp/workbench/hamm/service/CostService.scala:
```diff
@@ -41,5 +41,6 @@ class CostService(samAuthProvider: SamAuthProvider, dbRef: DbReference, jobTable
 
-final case class WorkflowCostResponse(workflowId: WorkflowId, cost: Double)
-final case class JobCostResponse(jobId: JobId, cost: Double)
+trait ServiceResponse
```
What's the point of having this `trait`?

ansingh7115 replied:
Oops - left that in after deleting a thing I was trying

ansingh7115 commented on server/src/test/scala/org/broadinstitute/dsde/workbench/hamm/api/HammRoutesSpec.scala:
```diff
@@ -0,0 +1,104 @@
+  override def beforeEach() = {
+    mockWorkflowTable.workflows += TestData.testWorkflow
+    mockJobTable.jobs += TestData.testJob
+
```
Most likely it was my elbow on the space bar :P

Qi77Qi commented on core/src/main/scala/org/broadinstitute/dsp/workbench/hamm/db/WorkflowTable.scala:
```diff
@@ -0,0 +1,133 @@
+    cost: Double)
+
+final case class Label(key: String, value: String)
+//final case class WorkflowCost(workflowId: WorkflowId, cost: Double)
```
can you remove this line

Qi77Qi commented on core/src/main/scala/org/broadinstitute/dsp/workbench/hamm/db/WorkflowTable.scala:
```diff
@@ -0,0 +1,133 @@
+  }
+
+
+  //ToDo: Make this better
```
when are you going to make this better :)

ansingh7115 replied:
something tells me never

Qi77Qi commented on core/src/main/scala/org/broadinstitute/dsp/workbench/hamm/db/db.scala:
```diff
@@ -0,0 +1,19 @@
+  val endTimeFieldName = "END_TIME"
+  val labelsFieldName = "LABELS"
+  val costFieldName = "COST"
+  val callNameFieldName = "CALL_Name"
```
call_fqn

Qi77Qi commented on core/src/test/scala/org/broadinstitute/dsp/workbench/hamm/TestExecutionContext.scala:
```diff
@@ -0,0 +1,20 @@
+  implicit val testExecutionContext = new TestExecutionContext()
+}
+
+class TestExecutionContext() extends ExecutionContextExecutor {
```
why does this need to be a separate class?

ansingh7115 replied:
I was told it was good practice - this allows people to instantiate a separate TestExecutionContext if they want

Qi77Qi commented on project/Dependencies.scala:
```diff
@@ -1,28 +1,43 @@
+    "io.sentry"                         %  "sentry-logback"      % "1.7.16",          // see doc https://docs.sentry.io/clients/java/modules/logback/
+    "org.broadinstitute.dsde.workbench" %% "sam-client"          % samV,
+    "org.http4s"                        %% "http4s-blaze-client" % http4sVersion,
+    "org.http4s"                        %% "http4s-blaze-server" % http4sVersion,
```
can you move this out of common...core module doesn't need `http4s-blaze-server`, neither automation
